src/util/http/UrlParser.cpp

```diff
@@ -5,114 +5,33 @@
-  std::string decoded;
-  for (size_t i = 0; i < url.size(); ++i) {
-    if (url[i] == '+') {
-      decoded += ' ';
-    } else if (url[i] == '%' && i + 2 < url.size()) {
-      char h1 = tolower(url[i + 1]);
-      if (h1 >= '0' && h1 <= '9') {
-        h1 = h1 - '0';
-      } else if (h1 >= 'a' && h1 <= 'f') {
-        h1 = h1 - 'a' + 10;
-      } else {
-        decoded += '%';
-        continue;
-      }
-      char h2 = tolower(url[i + 2]);
-      if (h2 >= '0' && h2 <= '9') {
-        h2 = h2 - '0';
-      } else if (h2 >= 'a' && h2 <= 'f') {
-        h2 = h2 - 'a' + 10;
-      } else {
-        decoded += '%';
-        continue;
-      }
-      decoded += static_cast<char>(h1 * 16 + h2);
-      i += 2;
-    } else {
-      decoded += url[i];
-    }
+UrlParser::ParsedHTTPRequest UrlParser::parseRequestTarget(
+    std::string_view target) {
+  UrlParser::ParsedHTTPRequest parsed;
+  auto result = boost::urls::parse_origin_form(target);
+  if (result.has_error()) {
+    AD_THROW(absl::StrCat("Failed to parse URL: \"", target, "\"."));
   }
-  return decoded;
-}
+  boost::url url = result.value();
 
-// ___________________________________________________________________________
-UrlParser::UrlPathAndParameters UrlParser::parseGetRequestTarget(
-    std::string_view target, bool urlDecode) {
-  UrlPathAndParameters result;
+  parsed.path_ = url.path();
+  parsed.parameters_ = paramsToMap(url.params());
 
-  // Remove everything after # (including it). Does nothing if there is no #.
-  // Don't do this is `urlDecode == false` because in that case, the given
-  // string contains an unencode SPARQL query, which frequently contains a # as
-  // a regular character.
-  if (urlDecode == true) {
-    target = target.substr(0, target.find('#'));
-  }
-
-  // Set `_path` and remove it from `target`. If there is no query string (part
-  // starting with "?"), we are done at this point.
-  size_t index = target.find('?');
-  result._path = target.substr(0, index);
-  if (index == std::string::npos) {
-    return result;
-  }
-  target.remove_prefix(index + 1);
+  return parsed;
+}
 
-  // Parse the query string and store the result in a hash map. Throw an error
-  // if the same key appears twice in the query string. Note that this excludes
-  // having two "cmd=..." parameters, although that would be meaningful (though
-  // not necessary) to support.
-  while (true) {
-    auto next = target.find('&');
-    auto paramAndValue =
-        parseSingleKeyValuePair(target.substr(0, next), urlDecode);
-    auto [iterator, isNewElement] =
-        result._parameters.insert(std::move(paramAndValue));
+// _____________________________________________________________________________
+ad_utility::HashMap<std::string, std::string> UrlParser::paramsToMap(
+    boost::urls::params_view params) {
+  ad_utility::HashMap<std::string, std::string> result;
+  for (auto param : params) {
+    auto [iterator, isNewElement] = result.insert({param.key, param.value});
```

Comment:
As this is not a programming error, it should not `AD_THROW` but simply throw a `std::runtime_error` or something similar. And you could also integrate both values of the duplicate key into the error message ( use `absl::StrCat` to concatenate strings).